src/security/PeerOptions.cc

```diff
@@ -297,130 +297,130 @@ static struct ssl_option {
 
 } ssl_options[] = {
 
-#if SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG
+#if defined(SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG)
```

Comment:
This and similar changes seem to contradict OpenSSL v3 code that does not (AFAICT) support these options but defines these constants:
https://github.com/openssl/openssl/blob/0fcf2351ecff5db21cba431704e4da631b74904a/include/openssl/ssl.h.in#L425-L433

This change confuses our supported option searching code in Security::PeerOptions::parseOptions() which assumes that options listed here have non-zero values. That code may still "work" after this change but that is essentially an accident that we should not rely on.

Could you please clarify why these preprocessor conditions should be changed?

Reply:
This is intended to solve two problems;
  1) compile warning/error with opensslv3 about some of the option definitions being non-constexpr, and
  2) backward compatibility. So that Squid built with libssl3 can take old config files using the deprecated options from libssl1.x

Reply:
> compile warning/error with opensslv3 about some of the option definitions being non-constexpr

Does this "compile warning/error" problem apply to the line this change request is attached to? It does not seem to apply (because `SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG` is defined to be `0x0` so the unchanged code should have compiled, but perhaps I am missing something.

If there are two classes of PR changes here (i.e., PR changes necessary to avoid "compile warning/error with opensslv3" and all others), then let's use this change request for the first class.

> backward compatibility. So that Squid built with libssl3 can take old config files using the deprecated options from libssl1.x

We do not want that kind of backward compatibility IMO: Squid should not silently swallow unsupported options it knows about. A configuration using `SSLREF2_REUSE_CERT_TYPE_BUG` should either stop working after the upgrade due to a fatal configuration error or, at the very least, log an error that the configured option is no longer supported.

Reply:
The options defined [1] to ```0x0``` are in the backward compatibility group. For slightly older libssl releases where they were/maybe set to a macro.

[1] the definition value is specific to the OpenSSL code you looked at. The const/macro nature and resulting value can and does change between releases of OpenSSL.

Reply:
> the definition value is specific to the OpenSSL code you looked at. The const/macro nature and resulting value can and does change between releases of OpenSSL.

I looked at more than 200 OpenSSL tagged releases (all I could find). All of them define `SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG`  to be a constant (`0x00000008L`, `0x0`, or `0x0U`). In OpenSSL v3, it is defined to be `0x0`. These facts support my change request: There are known cases where the proposed change makes things worse and no known cases where it makes things better. Thus, unless you have specific evidence to the contrary, this (and equivalent) changes should be undone.

If there are some similarly-looking PR changes that do not follow the above pattern of constant macro values, then they may need to be treated differently, of course, but that would be a discussion for another piece of code.

Reply:
Package: libssl-dev 3.0.3-5
```
In file included from ../../../compat/openssl.h:44,
                 from ../../../src/security/Context.h:15,
                 from ../../../src/security/forward.h:14,
                 from ../../../src/security/PeerOptions.h:14,
                 from ../../../src/security/PeerOptions.cc:16:
../../../src/security/PeerOptions.cc:330:5: error: missing binary operator before token "1"
  330 | #if SSL_OP_TLS_ROLLBACK_BUG
      |     ^~~~~~~~~~~~~~~~~~~~~~~
../../../src/security/PeerOptions.cc:335:5: error: missing binary operator before token "1"
  335 | #if SSL_OP_ALL
      |     ^~~~~~~~~~
../../../src/security/PeerOptions.cc:370:5: error: missing binary operator before token "1"
  370 | #if SSL_OP_CIPHER_SERVER_PREFERENCE
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../src/security/PeerOptions.cc:380:5: error: missing binary operator before token "1"
  380 | #if SSL_OP_NO_SSLv3
      |     ^~~~~~~~~~~~~~~
../../../src/security/PeerOptions.cc:385:5: error: missing binary operator before token "1"
  385 | #if SSL_OP_NO_TLSv1
      |     ^~~~~~~~~~~~~~~
../../../src/security/PeerOptions.cc:392:5: error: missing binary operator before token "1"
  392 | #if SSL_OP_NO_TLSv1_1
      |     ^~~~~~~~~~~~~~~~~
../../../src/security/PeerOptions.cc:399:5: error: missing binary operator before token "1"
  399 | #if SSL_OP_NO_TLSv1_2
      |     ^~~~~~~~~~~~~~~~~
../../../src/security/PeerOptions.cc:406:5: error: missing binary operator before token "1"
  406 | #if SSL_OP_NO_TLSv1_3
      |     ^~~~~~~~~~~~~~~~~
../../../src/security/PeerOptions.cc:413:5: error: missing binary operator before token "1"
  413 | #if SSL_OP_NO_COMPRESSION
      |     ^~~~~~~~~~~~~~~~~~~~~
../../../src/security/PeerOptions.cc:418:5: error: missing binary operator before token "1"
  418 | #if SSL_OP_NO_TICKET
      |     ^~~~~~~~~~~~~~~~
```

Reply:
> Treating some `SSL_OP_*` macros as `#if X` and others as `#if defined(X)` is inconsistent and odd looking code. It also increases our maintenance burden in future - requiring us to track OpenSSL (and its distributors) patching of libssl API definitions for these macros.

Finally we are starting to make progress, thank you!

```C
#define SSL_OP_BIT(n)  ((uint64_t)1 << (uint64_t)n)
...
# define SSL_OP_TLS_ROLLBACK_BUG                         SSL_OP_BIT(23)
...
# define SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG         0x0
```

So we have a group of macros that use constants (including `SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG` that triggered this change request) and a group of macros that use constant expressions (including `SSL_OP_TLS_ROLLBACK_BUG` in the error output you have now shared). A compiler does not allow us to process macros in both categories using `#if` conditionals. The logic represented by those macros does not allow us to process macros in both categories using `#ifdef` conditionals alone. A given macro may travel from one category to another (or might even disappear if we do not cut support for old OpenSSL versions soon enough). We want to be consistent and future-proof, so we need to develop a single solution that covers all two (or three) categories.

I suggest using this approach:

1. Use `#if defined(SSL_OP_X)` to test the existence of an OpenSSL `SSL_OP_X` macro (during preprocessing). The PR already contains all the necessary changes for this AFAICT.
2. Use `if (SSL_OP_X)` or equivalent to test whether the corresponding OpenSSL feature/option/etc. is actually enabled/supported by OpenSSL (during compilation). The current PR is missing those changes, but they are fairly easy to add.

For this specific change request, we need to adjust the ssl_options loop in Security::PeerOptions::parseOptions() so that it uses Optional for `value` (instead of misinterpreting zero `value` as "not found").

Reply:
> Finally we are starting to make progress, thank you!
> 

Does this mean you finally accept that I know what I'm doing here?



> I suggest using this approach:
> 
>     1. Use `#if defined(SSL_OP_X)` to test the existence of an OpenSSL `SSL_OP_X` macro (during preprocessing). The PR already contains all the necessary changes for this AFAICT.
> 

Indeed. The code you objected to has been doing this from the start.

>     2. Use `if (SSL_OP_X)` or equivalent to test whether the corresponding OpenSSL feature/option/etc. is actually enabled/supported by OpenSSL (during compilation). The current PR is missing those changes, but they are fairly easy to add.

The existing code in Security::PeerOptions::parseOptions() already implements this is a far simpler way than you are proposing.

Reply:
The OP request:
 > Could you please clarify why these preprocessor conditions should be changed?

Has been answered already several times.

As for the subsequent request to "use Optional for value". That is requesting a behavioural change in Squid unrelated to the bug being fixed by this PR. I am leaving it out of scope here.

Reply:
The current official code treats missing array entries as ```value=0``` and produces the ```"ERROR: Unknown TLS option```.

This PR allows 0-value to exist in the array. Regardless of that existence their parsed ```value=0``` property stays unchanged and they produce the ```"ERROR: Unknown TLS option```. Exactly as if they were not listed in the array at all.

AIUI, your request is to use Optional to perform some **new** error handling (eg different debugs message or throw) for the 0-value vs undefined option name. That is unnecessary.

Reply:
I do not see the issue you are ...

 * What you describe as a "new problem" is the same behaviour which has **always** occured on every Squid + OpenSSL upgrade. When value==0 for any option Squid rejects the squid.conf text. This is true regardless of how value became 0.

 * This PR is the implementation which **avoids** producing "new errors".

Can you please point me at the exact line sequence where you are seeing these alleged "new problems" ?

Reply:
> Can you please point me at the exact line sequence where you are seeing these alleged "new problems" ?

Sure, the following `if` statement assumes that `!value` means "not found" when, in fact, it no longer means that in the PR code. In the PR code `!value` means "either not found or found but not supported".

https://github.com/squid-cache/squid/blob/c25648fb4eca515f8a6a535cc93c2ac2eccc292b/src/security/PeerOptions.cc#L490-492

I did not check for other occurrences, but you should.

Reply:
Bit of a stretch there (none of the array options match hex numeric format). But okay, changed in [5a875b2](https://github.com/squid-cache/squid/pull/694/commits/5a875b21d2ae69d423239cdce0ba170a9c76671f)